plugins/plugin-java/che-plugin-java-plain/che-plugin-java-plain-server/src/main/java/org/eclipse/che/plugin/java/plain/server/inject/PlainJavaProjectSourceFolderWatcher.java

```diff
@@ -0,0 +1,197 @@
+    this.extensionService = extensionService;
+    this.projectManager = projectManager;
+    this.executorService =
+        Executors.newSingleThreadExecutor(
+            new ThreadFactoryBuilder()
+                .setNameFormat("WorkspaceUpdater-%d")
+                .setUncaughtExceptionHandler(LoggingUncaughtExceptionHandler.getInstance())
+                .setDaemon(true)
+                .build());
+  }
+
+  @PostConstruct
+  protected void startWatchers() {
+    int watcherId =
+        manager.registerByMatcher(
+            folderMatcher(),
+            s -> report(s, FileChangeType.Created),
+            s -> report(s, FileChangeType.Changed),
+            s -> report(s, FileChangeType.Deleted));
+
+    watcherIds.add(watcherId);
+    eventService.subscribe(this::onServerInitialized, LanguageServerInitializedEvent.class);
+    eventService.subscribe(this::onProjectUpdated, ProjectUpdatedEvent.class);
+  }
+
+  @PreDestroy
+  public void stopWatchers() {
+    watcherIds.stream().forEach(id -> manager.unRegisterByMatcher(id));
+  }
+
+  private void onServerInitialized(LanguageServerInitializedEvent event) {
+    LOG.debug(
+        "Registering source folder watching operations for language server : {}", event.getId());
+
+    String id = event.getId();
+    LanguageServer languageServer = event.getLanguageServer();
+    languageServers.add(languageServer);
+  }
+
+  private void onProjectUpdated(ProjectUpdatedEvent event) {
+    executorService.submit(
+        () -> {
+          setInitialPathsToWatch(prefixURI(event.getProjectPath()));
+        });
+  }
+
+  private PathMatcher folderMatcher() {
+    return it -> isSourceFolder(it);
+  }
+
+  private void report(String path, FileChangeType changeType) {
+    languageServers.stream().forEach(ls -> send(ls, path, changeType));
+  }
+
+  private void send(LanguageServer server, String path, FileChangeType changeType) {
+    RegisteredProject project = projectManager.getClosestOrNull(path);
+    if (project == null) {
+      return;
+    }
+    DidChangeWatchedFilesParams params =
+        new DidChangeWatchedFilesParams(
+            Collections.singletonList(new FileEvent(prefixURI(path), changeType)));
+    server.getWorkspaceService().didChangeWatchedFiles(params);
+  }
+
+  private boolean isSourceFolder(Path path) {
+    if (!isDirectory(path)) {
+      return false;
+    }
+
+    String wsPath = removePrefixUri(path.toUri().toString());
+    RegisteredProject project = projectManager.getClosestOrNull(wsPath);
+    if (project == null) {
+      return false;
+    }
+
+    return (getSourceFolders(project.getPath())
+            .stream()
+            .filter(p -> Paths.get(wsPath).startsWith(Paths.get(p)))
+            .count()
+        > 0);
+  }
+
+  private void setInitialPathsToWatch(String projectUri) {
+    List<String> sourceFolderLocations;
+    try {
+      sourceFolderLocations = extensionService.getAllSourceFoldersLocations(projectUri);
```

Comment:
We should not use a synchronous call here, rather use the promise we get from "WorkspaceService.executeCommand()". Then we can get rid of the executor service.

Reply:
Re-implemented